Pal 867 resource service auditing (#231)
gchq/Palisade-services: resource-service/README.md

(fields marked with * are acquired from headers metadata)

The service accepts a `ResourceRequest` from the User service, finds all the resources associated with the resourceId as a stream, the stream is 

> then consumed and for each resource a `ResourceResponse` is created is then sent onto the Policy service for further processing.

Reviewer: ```suggestion
then consumed, and for each resource a `ResourceResponse` is created and then sent to the Policy service for further processing.
```

Author: Updated

## Kafka Interface

The application receives 3 messages for each token, a `START` message, a message containing a `ResourceRequest` and an `END` message. The `START` gets 

> consumed by the service but no action taken on it and it gets written to the `resource` Kafka topic. The `ResourceRequest` message then gets consumed 

Reviewer: ```suggestion
consumed by the service, it is then acknowledged as the start of the resources and is then written to the `resource` Kafka topic. The `ResourceRequest` message then gets consumed 
```

Author: Updated

by the service and for each resource a `ResourceResponse` object is created. This then gets written to the `resource` Kafka topic. Once all the 

> `ResourceResponse`s have been written to the topic the `END` message gets written to the `resource` topic to mark the end of the resources for this 

Reviewer: ```suggestion
`ResourceResponse`s have been written to the topic, the `END` message gets written to the `resource` topic to mark the end of the resources for this 
```

Author: Updated

> request. In case any errors are encountered, the original request, along with the thrown exception are both captured in an 

Reviewer: ```suggestion
request. If any errors are thrown within the service, the original request, along with the thrown exception are captured in an 
```

Author: Updated

```

## Uploading resources to the backing store on service start-up

It may be that some example resources may need to be added to the backing store before, for example, a test run of the Palisade system gets performed. This is solved by 
using Spring to upload resource(s) to the service from a yaml file. An example of this can be seen in this
[testresource.yaml](src/contract-tests/resources/application-testresource.yaml) file which adds the resource information to the backing store when the service starts up.

## Hadoop and Windows

Windows users will have problems with Hadoop integration testing. Included [here](../../Palisade-services/resource-service/src/contract-tests/test/resources/hadoop-3.2.1) is a Windows-compatible set of hadoop binaries.

To 'install' on Windows, an additional step is required - copy the [hadoop.dll](../../Palisade-services/resource-service/src/contract-tests/test/resources/hadoop-3.2.1/bin/hadoop.dll) to `C:\Windows\System32`. This should enable

> the `HadoopResourceServiceTest` to run correctly.

Reviewer: The install location for hadoop is wrong. 
The hadoop resources are currently in component tests/resources/hadoop
not, resource-service/src/contact-tests/test? /resources

Author: Updated

|  resource: JSON         | --------------------------- ---------------------------
---------------------------
```

(`*` marks an indexable field)

### Motivation

To ease load on an external resource-service provider, a cache-like storage mechanism is used to return data for previously-queried resources. The code for this can be found in
the [StreamingResourceServiceProxy](src/main/java/uk/gov/gchq/palisade/service/resource/service/StreamingResourceServiceProxy.java) and [JpaPersistenceLayer](src/main/java/uk/gov/gchq/palisade/service/resource/repository/JpaPersistenceLayer.java).

> Due to the tree-structure of filesystems, naïve caching of requests-responses results in a large amount of data while not providing any "smart" methods to the caching mechanism. For example, requesting `/some/directory/with-lots-of-files` may return (and

Reviewer: Is naive the right word here? should it be native?

Author: Updated
